### PR TITLE
front: save NGE node positions in localStorage

### PR DESCRIPTION
**Pull request by emersion**
Short-term fix for setup persistance for node positions in localStorage

See long-term solution : https://github.com/osrd-project/osrd-confidential/issues/597

### CODE REVIEW

**Review by louisgreiner — APPROVED** [review]
First version of node position persistance, to allow user to work on its browser with NGE.

Remains to do : 
- [ ] save position at node creation (for now it only works after going back to Micro then back to Macro)
- [ ] handle non-trigram nodes

These enhancements can be done in a later work. Howerver, the current feature is highly needed for user friendliness and a good fix for now.

Later, this persistance layer will be implemented in the database, see https://github.com/osrd-project/osrd-confidential/issues/597

**Comment by louisgreiner** [comment]
Bugs noticed while testing:
- entries in `localStorage` are not removed when a node disappear while Macro->Micro. For example, if 2 nodes are connected by only 1 trainrun and that we delete the trainrun, the nodes will disappear after doing Macro->Micro, but not their entries in `localStorage`
- when editing the `Operational point` name, an update event is sent at EVERY change, and thus creating a new entry for EACH in `localStorage`. For example, if you enter "PLY", you'll have an entry for "P", "PL", and "PLY". Worst case scenario : you already have an entry for "PL", which ends up being rewrite with the position of "PLY". **EDIT**: fixed by adding a debounce time of 2000ms, but the long term fix of that should be on NGE side, and then revert this one

File: front/src/applications/operationalStudies/components/MacroEditor/types.ts [hunk 1]
```diff
@@ -154,8 +154,9 @@ export type NetzgrafikDto = {
...
 };
 
-export type NGETrainrunEvent = {
+export type NGEEvent = {
   type: 'create' | 'delete' | 'update';
-  objectType: 'trainrun';
-  trainrun: Trainrun;
+  objectType: 'node' | 'trainrun';
+  trainrun?: Trainrun;
+  node?: Node;
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/types.ts]: Use a type like this, this way we know in the type that if the objectType is node, then we have a Node and same for the trainrun
```suggestion
  } & (
  | {
      objectType: 'trainrun';
      trainrun: Trainrun;
    }
  | { objectType: 'node'; node: Node }
);
```

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts [hunk 2]
```diff
@@ -113,4 +114,59 @@ const handleOperation = async ({
...
+}: {
+  event: NGEEvent;
+  dispatch: AppDispatch;
+  timeTableId: number;
+  netzgrafikDto: NetzgrafikDto;
+}) => {
+  const { type } = event;
+  switch (event.objectType) {
+    case 'node':
+      handleNodeOperation({ type, node: event.node!, timeTableId });
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts]: Once you have modified the NGEEvent, you can remove the `!` after `node` l156 and `trainrun` line161

Comment (louisgreiner) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts]: true !

File: front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    entries = JSON.parse(rawNodes).map((node: Node) => [node.trigram, node]);
+  }
+
+  nodesCache = new Map(entries);
+  return nodesCache;
+};
+
+const saveNodes = (nodes: Map<string, Node>) => {
+  localStorage.setItem(NODES_LOCAL_STORAGE_KEY, JSON.stringify([...nodes.values()]));
+};
```
Comment (Uriel-Sautron) [inline front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts]: When the page is refreshed, our saved nodes no longer have their saved positions.

```suggestion
const loadNodes = () => {
  if (nodesCache) return nodesCache;

  let entries = [];
  const rawNodes = localStorage.getItem(NODES_LOCAL_STORAGE_KEY);
  if (rawNodes) {
    entries = JSON.parse(rawNodes).map((node: Node) =>  node);
  }

  nodesCache = new Map(entries);
  return nodesCache;
};

const saveNodes = (nodes: Map<string, Node>) => {
  localStorage.setItem(NODES_LOCAL_STORAGE_KEY, JSON.stringify([...nodes]));
};
```

Comment (louisgreiner) [reply front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts]: done thanks

**Review by Uriel-Sautron — COMMENTED** [review]
Tested, just a comment, otherwise good work

**Review by Uriel-Sautron — APPROVED** [review]

File: front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+let nodesCache: Map<string, Node> | null;
+
+const loadNodes = () => {
+  if (nodesCache) return nodesCache;
+
+  let entries = [];
+  const rawNodes = localStorage.getItem(NODES_LOCAL_STORAGE_KEY);
+  if (rawNodes) {
+    entries = JSON.parse(rawNodes).map((node: Node) => node);
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts:17]:
```suggestion
    entries = JSON.parse(rawNodes);
```

File: front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  nodesCache = new Map(entries);
+  return nodesCache;
+};
+
+const saveNodes = (nodes: Map<string, Node>) => {
+  localStorage.setItem(NODES_LOCAL_STORAGE_KEY, JSON.stringify([...nodes]));
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts:25]:
```suggestion
  localStorage.setItem(NODES_LOCAL_STORAGE_KEY, JSON.stringify(nodes));
```

File: front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts (hunk 4 above)
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/nodeStore.ts]:
```suggestion
    entries = JSON.parse(rawNodes).map((node: [string, Node]) => node);
```

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts [hunk 6]
```diff
@@ -113,4 +116,63 @@ const handleOperation = async ({
...
+}) => {
+  const { type } = event;
+  switch (event.objectType) {
+    case 'node':
+      handleNodeOperation({ type, node: event.node, timeTableId });
+      break;
+    case 'trainrun':
+      await handleTrainrunOperation({
+        type,
+        trainrun: event.trainrun!,
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts]: You can remove the ! here
```suggestion
        trainrun: event.trainrun,
```

**Review by clarani — APPROVED** [review]
LGTM, tested :white_check_mark: 
You can resolve the 2 remaining comments on your own :)

**Comment by emersion** [comment]
Thanks for finishing this up! Next time, please squash commits together.

**Pull request merged**